crates/harp/src/srcref.rs

```diff
@@ -0,0 +1,152 @@
+// attributes.
+impl TryFrom<RObject> for Vec<SrcRef> {
+    type Error = anyhow::Error;
+
+    fn try_from(value: RObject) -> anyhow::Result<Self> {
+        let srcref = unwrap!(value.attr("srcref"), None => {
+            return Err(anyhow!("Can't find `srcref` attribute"));
+        });
+
+        crate::List::new(srcref.sexp)
+            .iter()
+            .map(|x| SrcRef::try_from(RObject::view(x)))
+            .collect()
+    }
+}
+
+// Takes individual `srcref` attribute as input
+impl TryFrom<RObject> for SrcRef {
+    type Error = anyhow::Error;
+
+    fn try_from(value: RObject) -> anyhow::Result<Self> {
+        crate::r_assert_type(value.sexp, &[libr::INTSXP])?;
+        crate::r_assert_capacity(value.sexp, 6)?;
+
+        let value = unsafe { IntegerVector::new(value)? };
+
+        let line = std::ops::Range {
+            start: value.get_value(0)? - 1,
+            end: value.get_value(2)? - 1,
+        };
+
+        let line_parsed = if unsafe { value.len() >= 8 } {
+            std::ops::Range {
+                start: value.get_value(6)? - 1,
+                end: value.get_value(7)? - 1,
+            }
+        } else {
+            line.clone()
+        };
+
+        let column = std::ops::Range {
+            start: value.get_value(4)? - 1,
+            end: value.get_value(5)? - 1,
+        };
+
+        let column_byte = std::ops::Range {
+            start: value.get_value(1)? - 1,
+            end: value.get_value(3)? - 1,
+        };
+
+        Ok(Self {
+            line,
+            line_parsed,
+            column,
+            column_byte,
+        })
+    }
+}
+
+/// Creates the same sort of srcfile object as with `parse(text = )`.
+/// Takes code as an R string containing newlines, or as a R vector of lines.
+pub fn new_srcfile_virtual(code: SEXP) -> crate::Result<RObject> {
+    RFunction::new("base", "srcfilecopy")
+        .param("filename", "<text>")
+        .param("lines", code)
+        .call()
+}
+
+#[cfg(test)]
+mod tests {
+    use std::ops::Range;
+
+    use crate::test::r_test;
+    use crate::SrcRef;
+
+    #[test]
+    fn test_srcref() {
+        r_test(|| {
+            let exprs = crate::parse_exprs_with_srcrefs("foo\n\nś\nbar(\n\n)").unwrap();
+            let srcrefs: Vec<SrcRef> = exprs.try_into().unwrap();
+            let foo = &srcrefs[0];
+            let utf8 = &srcrefs[1];
+            let bar = &srcrefs[2];
+
+            assert_eq!(foo.line, Range { start: 0, end: 0 });
+            assert_eq!(foo.line_parsed, Range { start: 0, end: 0 });
+            assert_eq!(foo.column, Range { start: 0, end: 2 });
+            assert_eq!(foo.column_byte, Range { start: 0, end: 2 });
+
+            // `column_byte` is different because the character takes up two bytes
+            assert_eq!(utf8.line, Range { start: 2, end: 2 });
+            assert_eq!(utf8.line_parsed, Range { start: 2, end: 2 });
+            assert_eq!(utf8.column, Range { start: 0, end: 0 });
```

Comment:
I wonder if we should bump the column-end fields by 1. These empty ranges don't make much sense to me.